src/soar_sdk/connector.py

```
@@ -33,11 +61,204 @@ def __init__(self, actions_provider: "ActionsProvider") -> None:
+        response.raise_for_status()
+        return response
+
+    def put(self, endpoint: str, **kwargs: dict) -> httpx.Response:
+        """
+        Perform a GET request to the specfic endpoint using the soar client
+        """
+        filtered_kwargs = self.filter_keys(POST_VALID_KEYS, **kwargs)
+        headers = filtered_kwargs.pop("headers", {})
+        headers.update({"Referer": f"{self._client.base_url}/{endpoint}"})
+        response = self._client.put(endpoint, headers=headers, **filtered_kwargs)
+        response.raise_for_status()
+        return response
+
+    def _save_artifact(self, artifact: dict) -> tuple[bool, str, Optional[int]]:
+        artifact.update(
+            {k: v for k, v in self._artifact_common.items() if (not artifact.get(k))}
+        )
+        try:
+            json.dumps(artifact)
+        except TypeError as e:
+            exception = getattr(e, "message", str(e))
+            error_msg = f"json.dumps during Save artifact failed. Possibly a value in the artifact dictionary is not encoded properly. Exception: {exception}"
+            raise ActionFailure(error_msg) from e
+
+        if self.csrf_token:
+            try:
+                response = self.post("rest/artifact", json=artifact)
+            except Exception as e:
+                error_msg = f"Failed to add artifact: {e}"
+                raise ActionFailure(error_msg) from e
+
+            resp_data = response.json()
+
+            if "id" in resp_data:
+                return (True, "Artifact added successfully", resp_data["id"])
+
+            if "existing_artifact_id" in resp_data:
+                return (
+                    True,
+                    "Artifact already exists",
+                    resp_data["existing_artifact_id"],
+                )
+
+            msg_cause = resp_data.get("message", "NONE_GIVEN")
+            message = f"artifact addition failed, reason from server: {msg_cause}"
+            raise ActionFailure(message)
```

Discussion: should we automatically fail the action here, or should we give the caller a chance to react to this failure? Maybe we should have a subclass of `ActionFailure` to raise here? What are people's thoughts?

---

I think a subclass along the lines of SOARRestError could work. We wouldn't swallow this exception like we do with ActionFailure and would give the chance for the dev to handle this on their own